code/datums/components/surgery_initiator.dm

```diff
@@ -0,0 +1,281 @@
+	// make sure it's available to be done
+	if(forced_surgery)
+		for(var/datum/surgery/S in available_surgeries)
+			if(istype(S, forced_surgery))
+				procedure = S
+				break
+	else
+		procedure = input("Begin which procedure?", "Surgery", null, null) as null|anything in available_surgeries
+
+	if(!procedure)
+		return
+
+	return try_choose_surgery(user, target, procedure)
+
+/datum/component/surgery_initiator/proc/get_available_surgeries(mob/user, mob/living/target)
+	var/list/available_surgeries = list()
+
+	var/mob/living/carbon/carbon_target
+	var/obj/item/organ/external/affecting
+	if(iscarbon(target))
+		carbon_target = target
+		affecting = carbon_target.get_organ(check_zone(user.zone_selected))
+
+	for(var/datum/surgery/surgery as anything in GLOB.surgeries_list)
+		if(surgery.abstract)  // no choosing abstract surgeries
+			continue
+		if(!surgery.possible_locs.Find(user.zone_selected))
+			continue
+		if(affecting)
+			if(!surgery.requires_bodypart)
+				continue
+			if((surgery.requires_organic_bodypart && affecting.is_robotic()) || (!surgery.requires_organic_bodypart && !affecting.is_robotic()))
+				continue
+			if(surgery.requires_real_bodypart && !affecting.is_primary_organ())
+				continue
+		else if(carbon_target && surgery.requires_bodypart) //mob with no limb in surgery zone when we need a limb
+			continue
+		if(surgery.lying_required && target.body_position != LYING_DOWN)
+			continue
+		if(!surgery.self_operable && target == user)
+			continue
+		if(!surgery.can_start(user, target))
+			continue
+		for(var/path in surgery.target_mobtypes)
+			if(istype(target, path))
+				available_surgeries += surgery
+				break
+
+	return available_surgeries
+
+/// Does the surgery de-initiation.
+/datum/component/surgery_initiator/proc/attempt_cancel_surgery(datum/surgery/the_surgery, mob/living/patient, mob/user)
+	var/selected_zone = user.zone_selected
+	/// We haven't even started yet. Any surgery can be cancelled at this point.
+	if(the_surgery.step_number == 1)
+		patient.surgeries -= the_surgery
+		user.visible_message(
+			"<span class='notice'>[user] draws [parent] away from [patient]'s [parse_zone(selected_zone)].</span>",
+			"<span class='notice'>You remove [parent] from [patient]'s [parse_zone(selected_zone)].</span>",
+		)
+
+		qdel(the_surgery)
+		return
+
+	if(!the_surgery.can_cancel)
+		return
+
+	// Don't make a forced surgery implement cancel a surgery.
+	if(istype(the_surgery, forced_surgery))
+		return
+
+	var/obj/item/close_tool
+	var/obj/item/other_hand = user.get_inactive_hand()
+
+	var/is_robotic = !the_surgery.requires_organic_bodypart
+	var/datum/surgery_step/close_carbon = new /datum/surgery_step/generic/cauterize/premature()
+	var/datum/surgery_step/close_robo = new /datum/surgery_step/robotics/external/close_hatch/premature()
+	var/datum/surgery_step/chosen_close_step
+
+	if(is_robotic)
+		chosen_close_step = close_robo
+	else
+		chosen_close_step = close_carbon
```

Comment:
why new them when we might not use them? you dont need the close_robo and close_carbon vars.
just do `var/datum/surgery_step/close_step = null` then new them in the if else